cli/deployments/traefik.go

```diff
@@ -45,8 +47,32 @@ func (k Traefik) Describe() string {
 	return emoji.Sprintf(":cloud:Traefik version: %s\n:clipboard:Traefik Ingress chart: %s", traefikVersion, traefikChartURL)
 }
 
+// Delete removes traefik from kubernetes cluster
 func (k Traefik) Delete(c kubernetes.Cluster) error {
-	return c.Kubectl.CoreV1().Namespaces().Delete(context.Background(), traefikDeploymentID, metav1.DeleteOptions{})
+	message := "Deleting Traefik"
+	s := spinner.New(spinner.CharSets[11], 100*time.Millisecond)
+	s.Suffix = emoji.Sprintf(" %s :zzz:", message)
+	s.Start()
+
+	currentdir, err := os.Getwd()
+	if err != nil {
+		return errors.New("Failed uninstalling Traefik: " + err.Error())
+	}
+
+	helmCmd := fmt.Sprintf("helm uninstall traefik --namespace '%s'", traefikDeploymentID)
```

Comment:
This one is tricky. If we find traefik already running on the cluster [we simply skip the installation](https://github.com/SUSE/carrier/blob/main/cli/deployments/traefik.go#L106) (as it happens on k3d for example).

We need a way to mark resources that belong to Carrier so we can decide whether to delete them or not when uninstalling. Maybe we can use a label on the namespace? There are other ways to achieve the same but I can't think of any better.

Reply:
:+1: on that being tricky and needing a more careful approach. Well, I will see what happens on the k3s cluster I use. IIRC rohit uses minikube, where traefik is installed by carrier, so the issue would not be seen.

Reply:
Other ways: Maybe a carrier-specific config map which lists the namespaces/resources carrier created/has ownership of ?  ... No, on further thought I do not like that. It has the information separate from the resources itself, allowing for desync. Labels on the namespace might really be best. (I do not see an easy way to label all the resources in the namespace, and it might be overkill also).

Reply:
Rohit already suggested to tag the resources and delete them but we decided against it to keep it simple for now.
The point does make sense, the question though is if we really want to support multi-tenants, because atm all namespaces and similar seem to be static and would potential conflict in such an environment.

Reply:
Rohit already suggested to tag the resources and delete them but we decided against it to keep it simple for now.
The point does make sense, the question though is if we really want to support multi-tenants, because atm all namespaces and similar seem to be static and would potential conflict in such an environment.

Reply:
Right now, the only component we are re-using if available is traefik. In the future we may decide to do the same with tekton or others. In any case, if we re-use something which exists, we shouldn't try to delete it on uninstall. To make sure we don't accidentally re-use an existing namespace, we could prefix all the namespaces we create with `carrier-` or something similar.

Reply:
This is not about mutli-tenant. Ir is about differences in the foundations.

For a minikube cluster carrier installs traefik.
For a k3s cluster traefik comes pre-installed.

Trying to uninstall traefik there may break cluster.

Actually not true, as I just reached step 7 in my plan (above) and got the following error:

```
Carrier uninstalling...
⚓ Detected kubernetes platform: k3s
🌎 ExternalIPs: [192.168.176.2]
➰ InternalIPs: [192.168.176.2]
⣾ Deleting Traefik 💤  Couldn't install carrier: Failed uninstalling Traefik: Error: uninstall: Release not loaded: traefik: release: not found


real	0m0.177s
user	0m0.151s
sys	0m0.029s
```

I.e. the attempt to uninstall traefik ran into the fact that carrier had not installed it, so there was no proper helm release to act on.

Reply:
Ok. I disabled the `traefik` deployment in the uninstaller and redid things. Result:

```
arrier uninstalling...
⚓ Detected kubernetes platform: k3s
🌎 ExternalIPs: [192.168.176.2]
➰ InternalIPs: [192.168.176.2]
✔️  Quarks removed
✔️  Gitea removed
✔️  Eirini removed
✔️  Registry removed
⣾ Deleting Tekton 💤  Couldn't install carrier: Failed uninstalling Quarks: namespaces "tekton" not found
```

So there is something bogus in the tekton uninstall as well.

Further, the deletion of the eirini deployment killed the workload namespace as well, and thus will kill the deployed apps too. (In my case a failed deployed stuck in image pull for the sample app - That part is independent of the uninstall itself).

Should also fix the message of `Couldn't install carrier` to `Couldn't uninstall carrier`.

Reply:
Message fixes done with commit 15444437e5b77dc153136bac370ec5d0495202c8

Reply:
Yes, Namespace labelling makes sense. Will implement it.